### ISSUE TITLE
Autogenerate unique id column from PostGIS view if not exists (or at least be more descriptive upon failure)

### ISSUE THREAD

**Issue by iboates** [qgis/QGIS#46444]
### Feature description

I was trying to add a PostGIS view as a layer and it failed. After some research and experimentation, I found that it was failing because there was no column to be used as a unique id. So I simply added a `ROW_NUMBER() OVER ()` to the view definition and it loaded into QGIS without issue. There was no clear explanation of this requirement anywhere that I could see, it simply failed with an empty "reason" string.

Since views are inherently temporary by nature (I guess excluding the concept of materialized views), I suggest that QGIS simply generate its own id column inside the layer dynamically when this happens.

Here are my QGIS specs:

```
<!--StartFragment--><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>


QGIS version | 3.10.7-A Coruña | QGIS code revision | 7b4ca4c8d0
-- | -- | -- | --
Compiled against Qt | 5.11.2 | Running against Qt | 5.11.2
Compiled against GDAL/OGR | 3.0.4 | Running against GDAL/OGR | 3.0.4
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.29.0 | Running against SQLite | 3.29.0
PostgreSQL Client Version | 11.5 | SpatiaLite Version | 4.3.0
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.8
Compiled against PROJ | 6.3.2 | Running against PROJ | Rel. 6.3.2, May 1st, 2020
OS Version | Windows 10 (10.0)
Active python plugins | loadthemall; openlayers_plugin; OSMDownloader; QRectangleCreator; quick_map_services; timemanager; db_manager; processing

</body></html><!--EndFragment-->
```

**Comment by m-kuhn** [comment]
Can you try with a currently supported version? 3.20 is EOL.

**Comment by kannes** [comment]
For the record as I ended up here looking for contemporary info on what QGIS requires re: unique IDs:

It seems like modern QGIS will automatically add such a clause if no unique id specified. E.g. it turned my

`SELECT * FROM "public"."addresses" LIMIT 10`
into
```
SELECT row_number() over () AS _uid0_, * FROM (SELECT * FROM "public"."addresses" LIMIT 10
) AS _subq_0_
```
automatically when I did not specify a "Column(s) with unique values" for a query layer.

Good riddance to the days past from which I remembered this issue :)
